rpc/src/client/transport/websocket.rs

```diff
@@ -0,0 +1,703 @@
+            Error::client_internal_error(format!(
+                "failed while waiting for WebSocket driver task to terminate: {}",
+                e
+            ))
+        })?
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct GenericJSONResponse(serde_json::Value);
+
+impl Response for GenericJSONResponse {}
+
+#[derive(Debug)]
+struct WebSocketSubscriptionDriver {
+    stream: WebSocketStream<TokioAdapter<TcpStream>>,
+    router: SubscriptionRouter,
+    cmd_rx: ChannelRx<WebSocketDriverCmd>,
+    terminate_rx: ChannelRx<TerminateSubscription>,
+}
+
+impl WebSocketSubscriptionDriver {
+    fn new(
+        stream: WebSocketStream<TokioAdapter<TcpStream>>,
+        cmd_rx: ChannelRx<WebSocketDriverCmd>,
+        terminate_rx: ChannelRx<TerminateSubscription>,
+    ) -> Self {
+        Self {
+            stream,
+            router: SubscriptionRouter::default(),
+            cmd_rx,
+            terminate_rx,
+        }
+    }
+
+    async fn run(mut self) -> Result<()> {
+        // TODO(thane): Should this loop initiate a keepalive (ping) to the
+        //              server on a regular basis?
+        loop {
+            tokio::select! {
+                Some(res) = self.stream.next() => match res {
+                    Ok(msg) => self.handle_incoming_msg(msg).await?,
+                    Err(e) => return Err(
+                        Error::websocket_error(
+                            format!("failed to read from WebSocket connection: {}", e),
+                        ),
+                    ),
+                },
+                Some(cmd) = self.cmd_rx.recv() => match cmd {
+                    WebSocketDriverCmd::Subscribe {
+                        req,
+                        event_tx,
+                        result_tx,
+                    } => self.subscribe(req, event_tx, result_tx).await?,
+                    WebSocketDriverCmd::Close => return self.close().await,
+                },
+                Some(term) = self.terminate_rx.recv() => self.unsubscribe(term).await?,
+            }
+        }
+    }
+
+    async fn send(&mut self, msg: Message) -> Result<()> {
+        self.stream.send(msg).await.map_err(|e| {
+            Error::websocket_error(format!("failed to write to WebSocket connection: {}", e))
+        })
+    }
+
+    async fn subscribe(
+        &mut self,
+        req: request::Wrapper<subscribe::Request>,
+        event_tx: ChannelTx<Result<Event>>,
+        mut result_tx: ChannelTx<Result<()>>,
+    ) -> Result<()> {
+        let id: SubscriptionId = req.id().clone().try_into()?;
+        let query = req.params().query.clone();
+        if let Err(e) = self
+            .send(Message::Text(serde_json::to_string_pretty(&req).unwrap()))
+            .await
+        {
+            let _ = result_tx.send(Err(e)).await;
+            return Ok(());
+        }
+        self.router
+            .pending_add(id.as_ref(), &id, query, event_tx, result_tx);
```

Comment:
This was technically an implementation of the [`AsRef` trait](https://doc.rust-lang.org/std/convert/trait.AsRef.html) to provide a `&str` reference, but looking at it now it's probably somewhat ambiguous. I've now implemented an `as_str()` method to return a `&str` instead, which I think should be clearer.

Reply:
But it's also a bit jarring that these two distinct Ids are actually the same. Is that intended? Maybe could be documented/clarified somewhere the subscription id will be drawn from the json-rpc id ?

Reply:
Originally I had them both be exactly the same, but then I ran into an issue where I needed to send unsubscribe requests with their own request ID (distinct from the subscription ID) to ensure that the response from the remote endpoint is, in fact, exclusively relevant to the unsubscribe request. Otherwise we may confuse an error meant for the subscription with one meant for the unsubscribe request.

So subscribe requests' IDs are the same as the subscription ID (to correlate incoming events, and possibly errors, relating to the subscription), but unsubscribe requests' IDs need to be distinct.